src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg, size_t max_size, std::chrono::milliseconds keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{std::move(keep_alive)} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(user.name);
+        cfg.scram_password.set_value(user.pass);
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+client_ptr kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    auto it_hash = inner_hash.find(user.name);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size) {
+            if (!_cache.empty()) {
+                inner_list.pop_back();
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", user.name);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(user.name, client);
+    } else {
+        // Otherwise the user has a client
+        vlog(plog.debug, "Reuse client for user {}", user.name);
+
+        // Grab the node located at that position
+        auto node = inner_hash.extract(it_hash);
+
+        // Update the timestamp and put it to the front
+        node.value().last_used = timestamped_user::clock::now();
+        client = node.value().client;
+        inner_list.insert(inner_list.begin(), std::move(node));
```

Comment:
Maybe it's worth writing a benchmark.
1. `.extract` && `.insert`
2. `.modify` && `.relocate`
3. `boost::intrusive_list` + `absl::flat_hash_map`.

Reply:
TLDR -- The node is **not** added to a list or a hash table but instead I am using a list interface to add the node back to the underlying container.

IIUC, you may be thinking that a boost `multi_index_conatiner` maintains individual containers. In this code that would be an separate linked list and a separate hash table.

Based on my reading that is **not** how `multi_index_container` works though. The following is the definition:
```
The Boost Multi-index Containers Library provides a class template named multi_index_container which enables the construction of containers maintaining one or more indices with different sorting and access semantics. Indices provide interfaces similar to those of STL containers, making using them familiar.
```

The key piece here is " Indices provide interfaces similar to those of STL containers" so there is **only 1** underlying container but multi_index_container allows us to use STL interfaces (called indexes) to manipulate that data.

So the node is **not** added to a list or a hash table but instead I am using a list interface to add the node back to the underlying container.

Reply:
> only 1 underlying container

i'm not sure this is possible. how would a single primitive container maintain an ordering and a hash-based indexing? i'm open to the possibility of a specialization being used here internally, but afaict the most likely thing is that boost is hiding the fact that there are two internal structures, and it's just not clear how it works.

not knowing how it works is ok up until we have to start thinking about iterator invalidation etc... which will probably happen.

Reply:
A better link explaining Boost Multiindex inner container
https://www.boost.org/doc/libs/1_80_0/libs/multi_index/doc/tutorial/index.html

I think the figure in the docs does a better job explaining it

Reply:
> not knowing how it works is ok up until we have to start thinking about iterator invalidation etc... which will probably happen.

Fair point. I'll research this more to see. What I'll do in the meantime, Monday, is switch to boost::intrusive_list + absl::flat_hash_map since it's more accepted.

Reply:
I usually reach for boost multi-index for this type of thing since keeping all indexes up to date is conceptually simpler as it's a single operation, and is exception safe.

[Iterator stability](https://www.boost.org/doc/libs/1_75_0/libs/multi_index/doc/tutorial/basics.html#complexity) is documented: "Iterator and reference validity is preserved in the face of insertions, even for replace and modify operations."

If the consensus is that maintaining multiple containers is easier to understand, I won't block the review on it.

Reply:
Resolving after an offline convo with @dotnwat that this discussion as more for learning purposes. There may be potential performance improvements in the future though.